### DISCUSSION TITLE
Problems using the compiler at Stack Auth

### DISCUSSION THREAD

**Discussion by N2D4** [reactwg/react-compiler#38] [Something went wrong]
First of all, we're super excited for a future with the compiler, and can't wait to support [Stack Auth](https://github.com/stack-auth/stack) with it.

However, in our client SDK, we have a lot of APIs that use promises. Code might look like this:

```typescript
const user = await getUser({ or: 'redirect' });
const connection = await user.getConnectedAccount('facebook', { or: 'redirect' });
const token = await connection.getAccessToken();
```

Since we target React natively, we provide hook-based counterparts:

```typescript
export default function MyClientComponent() {
  const user = useUser({ or: 'redirect' });
  const connection = user.useConnectedAccount('facebook', { or: 'redirect' });
  const token = connection.useAccessToken();

  return token;
}
```

This is a very powerful pattern. It lets us model large data structures without exposing our users to too much complexity; also, it means the hook-based SDK is consistent with the Promise-based one.

Notably, this pattern can align with the rules of React; `useConnectedAccount` is on the prototype of `User`, just called with different `this` arguments. We never need to actually pass the hook around as a value.

Unfortunately the new React compiler's static analysis throws an error:

```typescript
Hooks must be the same function on every render, but this value may change over time to a different function. See https://react.dev/reference/rules/react-calls-components-and-hooks#dont-dynamically-use-hooks
```

It's true that this could be the case (for some implementation of `useUser`), but also there's no way for us as library authors to tell the compiler that the hook is safe to call. I would say the most common case is that `useConnectedAccount(...)` is a function on the User prototype, and that it's always the same (with different `this` arguments).

Disabling the compiler on our side is not an option; this is code that our users write, and the errors will be thrown on their side.

**Comment by aleclarson** [comment]
What does the compiler output look like?

**Reply by N2D4** [reply]
[Playground](https://playground.react.dev/#N4Igzg9grgTgxgUxALhAgHgBwjALgAgBMEAzAQygBsCSoA7OXASwjvwFkBPAYUqYTq5uEALbY6A3AAoAlPmAAdNvjiswBKGAQx8AXnyaEAVS0wpwfDmT4A5DASEm9xjfwBfGQG4l+FWoKqdBKMLGz6hjAAdIbCQQiMDgCCcKr00jbkiABGEBAA1jYANPKWMNZ2Dk7xuK4e3sqB6vi4+QJ6fnEhrNFayYhgYAAqrXSy9T749riwbC15AvVuIG5AA):

> InvalidReact: Hooks must be the same function on every render, but this value may change over time to a different function. See https://react.dev/reference/rules/react-calls-components-and-hooks#dont-dynamically-use-hooks (3:3)

**Reply by suhaotian** [reply]
this means you shouldn’t write higher order Hooks:
https://react.dev/reference/rules/react-calls-components-and-hooks#dont-dynamically-mutate-a-hook

**Reply by N2D4** [reply]
Right, though in this case the returned hook is always the same (and not dynamically mutated). It would be useful if there was a way to achieve this.

**Reply by guillaumebrunerie** [reply]
Can you change the API a bit so that `useConnectedAccount` is an independent function instead of being a method? So that you would write the following instead:
```js
const connection = useConnectedAccount(user, 'facebook', { or: 'redirect' });
```

**Reply by N2D4** [reply]
It'd technically be possible, that would kinda defeat the purpose though, as having all available hooks on a single object is quite a nice DX that is otherwise unachievable

**Reply by tho-graf** [reply]
I just want to briefly jump into the discussion as well. We also have certain functionality, like feature toggles, implemented as a hook. Depending on whether we’re in a test or production environment, we inject either the real hook or a simple mock via context. So during a single execution, the hook is never changed. We ran into exactly the same error. The rule might be a bit too strict in this case, perhaps? In our situation, we simply disabled the compiler inside the hook using `use no memo`.

**Comment by aniruddhaadak80** [comment]
Hi there! Taking a look at your issue regarding `Problems using the compiler at Stack Auth` in the context of `reactwg/react-compiler`...

When working with the React Compiler (React Forget), this type of behavior usually stems from how the compiler analyzes hook dependencies and component boundaries. Make sure that:
1. Your code strictly adheres to the Rules of React. The compiler bails out or behaves unexpectedly when it detects rule violations.
2. In certain cases, explicitly defining memoization boundaries or isolating the hook/component into its own tightly-scoped file helps the Babel plugin successfully parse the AST.
3. Verify your ESLint `eslint-plugin-react-compiler` rules are green before compilation, as the compiler relies heavily on those guarantees.

Hope this helps clarify the expected boundaries! Let me know if you have a minimal repro I can test against.
